test/test_common/simulated_time_system_test.cc

```diff
@@ -62,6 +76,95 @@ TEST_F(SimulatedTimeSystemTest, AdvanceTimeAsync) {
+
+  // Verify order.
+  EXPECT_EQ("p012", output_);
+}
+
+// Timers that are scheduled to execute and but are disabled first do not trigger.
+TEST_F(SimulatedTimeSystemTest, TimerOrderAndDisableTimer) {
+  trackPrepareCalls();
+
+  // Create 3 timers. The first timer should disable the second, so it doesn't trigger.
+  addCustomTask(0, '0', [this]() { timers_[1]->disableTimer(); });
+  addTask(1, '1');
+  addTask(2, '2');
+  EXPECT_EQ(3, timers_.size());
+
+  // Expect timers to execute in order since the timers are scheduled at have different times and
+  // that timer 1 does not execute because it was disabled as part of 0's execution.
+  advanceMsAndLoop(5);
+  // Verify that timer 1 was skipped.
+  EXPECT_EQ("p02", output_);
+}
+
+// Capture behavior of timers which are rescheduled without being disabled first.
+TEST_F(SimulatedTimeSystemTest, TimerOrderAndRescheduleTimer) {
+  trackPrepareCalls();
+
+  // Reschedule timers 1, 2 and 4 without disabling first.
+  addCustomTask(0, '0', [this]() {
+    timers_[1]->enableTimer(std::chrono::milliseconds(0));
+    timers_[2]->enableTimer(std::chrono::milliseconds(100));
+    timers_[4]->enableTimer(std::chrono::milliseconds(0));
+  });
+  addTask(1, '1');
+  addTask(2, '2');
+  addTask(3, '3');
+  addTask(10000, '4', false);
+  EXPECT_EQ(5, timers_.size());
+
+  // Rescheduling timers that are already scheduled to run in the current event loop iteration has
+  // no effect if the time delta is 0. Expect timers 0, 1 and 3 to execute in the original order.
+  // Timer 4 runs as part of the first wakeup since its new schedule time has a delta of 0. Timer 2
+  // is delayed since it is rescheduled with a non-zero delta.
+  advanceMsAndLoop(5);
+  EXPECT_EQ("p0134", output_);
+
+  advanceMsAndLoop(100);
+  EXPECT_EQ("p0134p2", output_);
+}
+
+// Disable and re-enable timers that is already pending execution and verify that execution is
+// delayed.
+TEST_F(SimulatedTimeSystemTest, TimerOrderDisableAndRescheduleTimer) {
+  trackPrepareCalls();
+
+  // Disable and reschedule timers 1, 2 and 4 when timer 0 triggers.
+  addCustomTask(0, '0', [this]() {
+    timers_[1]->disableTimer();
+    timers_[1]->enableTimer(std::chrono::milliseconds(0));
+    timers_[2]->disableTimer();
+    timers_[2]->enableTimer(std::chrono::milliseconds(100));
+    timers_[4]->disableTimer();
+    timers_[4]->enableTimer(std::chrono::milliseconds(0));
+  });
+  addTask(1, '1');
+  addTask(2, '2');
+  addTask(3, '3');
+  addTask(10000, '4', false);
+  EXPECT_EQ(5, timers_.size());
+
+  // timer 0 is expected to run first and reschedule timers 1 and 2. Timer 3 should fire before
+  // timer 1 since timer 3's registration is unaffected. timer 1 runs in the same iteration
+  // because it is scheduled with zero delay. Timer 2 executes in a later iteration because it is
+  // re-enabled with a non-zero timeout.
+  advanceMsAndLoop(5);
+  EXPECT_EQ("p0314", output_);
+
+  advanceMsAndLoop(100);
+  EXPECT_EQ("p0314p2", output_);
```

Comment:
See TimerImplTest.TimerOrderDisableAndReschedule in https://github.com/envoyproxy/envoy/pull/11564 for a similar test using real time.